dig.go

```diff
@@ -168,6 +168,14 @@ type containerStore interface {
 	// Returns the providers that can produce values for the given group and
 	// type.
 	getGroupProviders(name string, t reflect.Type) []provider
+
+	// Adds the nodes to the list of failed nodes of the graph in the container
+	// and marks the corresponding constructor as failed.
+	failNodes(params []*dot.Param, id uintptr)
```

Comment:
I think it would be preferable to pass the `provider` here since the
`containerStore` deals in `provider`s. The underlying call to the graph can
deal in raw ids.

That theoretically lets you drop the `params` argument too because the
`failNodes` implementation can do `provider.ParamList().DotParam()`.

(We can add a separate `fail*` method for the case when `id` is 0.)

Reply:
We need to pass the specific `param` in since we want to colour only the node that was required. If `provider.ParamList().DotParam()` was used instead, every node in the constructor will be coloured. That is technically correct since every node in the constructor failed to build, but I think is less desirable since we want to be able to tell which nodes are causing the failure.

What do you think is the desired behaviour?

Reply:
Ah, that makes sense. This is fine as-is, then.